lsp/nls/src/linearization/mod.rs

```diff
@@ -24,6 +25,24 @@ pub mod building;
 pub mod completed;
 pub mod interface;
 
+// Cache of completed items
+static mut LIN_CACHE: Option<HashMap<FileId, Completed>> = None;
+
+fn update_lin_cache(file: FileId, value: Completed) {
+    unsafe {
+        match &mut LIN_CACHE {
+            Some(table) => {
+                table.insert(file, value);
+            }
+            None => {
+                let mut table = HashMap::new();
+                table.insert(file, value);
+                LIN_CACHE = Some(table);
+            }
+        }
+    }
+}
+
```

Comment:
Does the linearization cache really need to be `static mut`? As far as I can tell, it's only accessed in `impl`-methods on `AnalysisHost`. Could it instead be a field there? That would also save us from these `unsafe` blocks.

Reply:
I agree. `unsafe` seems like a code smell in this situation, which doesn't look like anything special requiring something as powerful.

Reply:
Yes, my initial solution was to have a `&HashMap<FileId, Completed>`  in the `Completed` data structure, but then I ran into lifetime scoping issues. I then decided to `clone` the `lin_cache`, but the `lin_cache` gets very large (contains `stdlib` and other opened files), so this significantly slowed down the LSP.  We don't need it to be accessible from `AnalysisHost` methods, we actually want it to be accessible from `impl`-methods on `Completed`

Reply:
See [here](https://github.com/tweag/nickel/pull/993/commits/a8dc7c6649968efd789da812b9189f161087d1cc)

Reply:
If the problem is just lifetime scoping and `Completed` should actually hold a reference to the linearization cache, could that be solved with `Rc`'s?

Reply:
After using an `Rc` we just have this unsafe code: 
```rust
  pub fn new(
        linearization: Vec<LinearizationItem<Resolved>>,
        id_to_index: HashMap<ItemId, usize>,
        lin_cache: &HashMap<FileId, Completed>,
    ) -> Self {
        Self {
            linearization,
            id_to_index,
            lin_cache: unsafe { Rc::from_raw(lin_cache) },
        }
    }

```